### PR TITLE
[serveless-1.25] changed func-deploy task to use kn from serverless operator

### PR DESCRIPTION
**Pull request by jrangelramos**
update of func-deploy.yaml task to use kn from serverless-operator to perform func deploy.

### CODE REVIEW

File: pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml [hunk 1]
```diff
@@ -24,7 +24,20 @@ spec:
       description: The workspace containing the function project
   steps:
   - name: func-deploy
-    image: "registry.redhat.io/openshift-serverless-1/kn-plugin-func-rhel8:latest"
+    image: "registry.access.redhat.com/ubi8/ubi"
```
Comment (matejvasek) [inline pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml:27]: Shouldn't we use digest?

Comment (matejvasek) [reply pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml:27]: and io registry?

Comment (cardil) [reply pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml:27]: - The `registry.redhat.io/ubi8/ubi` requires [registry tokens](https://catalog.redhat.com/software/containers/ubi8/ubi/5c359854d70cc534b3a3784e?container-tabs=gti&gti-tabs=registry-tokens),
- The `registry.access.redhat.com/ubi8/ubi` can be used [without authentication](https://catalog.redhat.com/software/containers/ubi8/ubi/5c359854d70cc534b3a3784e?container-tabs=gti&gti-tabs=unauthenticated)
 
Most likely, we should stick with the `registry.access.redhat.com/ubi8/ubi` variant.

About the digest... That is the question, in fact!

Using the `registry.access.redhat.com/ubi8/ubi:latest` image will work, and has an advantage, that the image is always up-to-date. This automatically fixes all upcoming CVEs. On the other hand, if we decide to use SHA hash here, we'll be required to release Serverless just to fix this image, by bumping its SHA. Cumbersome, isn't?

Furthermore, in the disconnected clusters, this image needs to be [mirrored to bastion host](https://docs.openshift.com/container-platform/4.11/installing/disconnected_install/installing-mirroring-disconnected.html) to work. 

If we use the SHA hash, mirroring is simple. We have to put that image reference to the Operator's related images list. If we use the `:latest`, or a tag like `:8.6`, we could add the SHA hash of that image to the Operator's related images list, at the time of release. When mirroring is being done, the image will be mirrored by the hash, but will contain all assigned tags. This means the bastion registry will know all those assigned tags, not only the hash. This surely will work with the custom tags like `:8.6`. I have serious doubts whether it will work with the `:latest` tag. The `:latest` tag is ephemeral, assigned by registry.

**TL;DR:** Because of the above, I would vote to use `registry.access.redhat.com/ubi8/ubi:8.6`. 

For regular clusters, the image will be automatically updated, mitigating CVEs. For disconnected install, the image will be mirrored with a version that had `:8.6` tag at a time of our release. After a while, it may be outdated. It's not a big deal, though. Users could easily push the current `:8.6` image to the bastion host manually at a later date:

```shell
$ podman pull registry.access.redhat.com/ubi8/ubi:8.6
$ podman tag registry.access.redhat.com/ubi8/ubi:8.6 bastion.acme.localnet:5000/ubi8/ubi:8.6
$ podman push bastion.acme.localnet:5000/ubi8/ubi:8.6
```

WDYT? @jrangelramos @matejvasek @lance

File: pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml [hunk 2]
```diff
@@ -24,7 +24,20 @@ spec:
...
     script: |
+      arch="amd64"
```
Comment (cardil) [inline pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml:29]:
```suggestion
      set -Eeuo pipefail
      arch="amd64"
```

See: http://redsymbol.net/articles/unofficial-bash-strict-mode

We use pipes in the script, so definitely the `pipeline` is required. The `-E` and `-e` options ensures error will end the script. The `-u` fail in case a unknown variable is referenced.

File: pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml [hunk 3]
```diff
@@ -24,7 +24,20 @@ spec:
...
+      machine=$(uname -m)
+      case ${machine} in
+          ppc|ppc64)
+              arch="ppc64le"
+              ;;
+          s390|s390x)
+              arch="s390x"
```
Comment (cardil) [inline pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml:36]: We should fail if machine is not recognized:

```suggestion
          s390|s390x)
              arch="s390x"
              ;;
          *)
              echo "Unsupported archiecture type: ${machine}" 2>&1
              exit 56
```

See: https://linuxize.com/post/bash-redirect-stderr-stdout/

**Review by cardil — CHANGES_REQUESTED** [review]
Some cleanups, and decision we need to make...

File: pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml (hunk 1 above)
Comment (dsimansk) [reply pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml:27]: Although I like the idea of direct download from Serverless operator. I wonder if we should stick with more simple solution and try to use `kn-client` image for this release.

From @cardil summary, use of `registry.access.redhat.com/ubi8/ubi:8.6` sounds good to me. 

I wonder though if Tekton Pipelines doesn't provide a base "run my script" image that we could reuse here. Therefore we doesn't have to add any additional ubi image to our S-O `relatedImages` spec.

File: pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml (hunk 3 above)
Comment (jrangelramos) [reply pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml:36]: just curious here.: why return "56" code, also why should I care about redirect echo standard error here to std out.

Comment (cardil) [reply pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml:36]: No particular reason for `56`. It needs to be `!= 0` to indicate failure. It's good to not be `1` as often people use, so you could differentiate errors just by retcode (don't confuse 1:1 relation here).

You should send logs & errors to stderr when scripting. It maybe is not significant in this case, but it's a good habit to have.

Comment (jrangelramos) [reply pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml:36]: I know it should be different that 0, just curios about the number choice 56 and yea anyway I think redirect on echo here does not apply. thanks!!

File: pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml [hunk 4]
```diff
@@ -24,7 +24,29 @@ spec:
       description: The workspace containing the function project
   steps:
   - name: func-deploy
-    image: "registry.redhat.io/openshift-serverless-1/kn-plugin-func-rhel8:latest"
+    image: "registry.access.redhat.com/ubi8/ubi:8.6"
```
Comment (Kaustubh-pande) [inline pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml:27]: @jrangelramos Just curious, this task is for the tests or customer will use this too? if this is for customers too then we need `registry.redhat.io` instead of `registry.access.redhat.com`.

Comment (jrangelramos) [reply pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml:27]: Hi @Kaustubh-pande , yes this is for our customer.

Comment (cardil) [reply pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml:27]: I think you are not right on this @Kaustubh-pande. This is something that customers will execute, right. But, the same is for Knative services or simple pods running on OpenShift. We even advise the in the docs to use ubi as a base image for their apps.

Comment (Kaustubh-pande) [reply pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml:27]: I guess it is recommended for customers use `registry.redhat.io`, `registry.access.redhat.com` to provide unauthenticated access, at some point in the future it will decommission as well 

- https://access.redhat.com/RegistryAuthentication#learning-more-about-red-hat-registry-9
- https://source.redhat.com/groups/public/sbr_containers/cee_containers_sbr_wiki/terms_based_authenticated_registry_faq

Maybe I was wrong here but it is good to verify once with Naina or rajesh.

Comment (jrangelramos) [reply pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml:27]: @nainaz ^^^

**Review by lance — COMMENTED** [review]
/lgtm but
/hold for @nainaz to answer the question of which image registry should be used

**Comment by cardil** [comment]
/hold 

See discussion https://coreos.slack.com/archives/CLMP7R2G2/p1665151241591059

**Comment by cardil** [comment]
@lance It's not just a decision which registry to use. One of them requires authentication, and will not work. The same as the client-kn image.

There is an alternative approach discussed in https://coreos.slack.com/archives/CLMP7R2G2/p1665151241591059 which may allow using productized images in Tekton.

**Review by lance — COMMENTED** [review]
This looks nice and clean. Looking forward to some of the parameterization optimizations and `deploy` image we were talking about in later releases.
/lgtm
/hold for others

File: pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml (hunk 4 above)
Comment (cardil) [reply pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml]: We came up with different, better solution, of using client image directly.

File: pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml [hunk 5]
```diff
@@ -24,7 +24,16 @@ spec:
...
   steps:
   - name: func-deploy
-    image: "registry.redhat.io/openshift-serverless-1/kn-plugin-func-rhel8:latest"
-    script: |
-      export FUNC_IMAGE="$(params.image)"
-      func deploy --verbose --build=false --push=false --path=$(params.path)
+    image: "registry.redhat.io/openshift-serverless-1/client-kn-rhel8@sha256:51e683fc3d668c39c6100512492e7fd7ca8f3f910886b4302960e825936d7cfb" 
+    env:
+      - name: FUNC_IMAGE
+        value: "$(params.image)"
```
Comment (cardil) [inline pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml:30]: What this `FUNC_IMAGE` is?

**Review by cardil — CHANGES_REQUESTED** [review]

File: pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml (hunk 5 above)
Comment (jrangelramos) [reply pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml:30]: that enviroment variable is required to inform func which actual function should be deployed.

**Review by cardil — APPROVED** [review]
/lgtm

**Comment by matejvasek** [comment]
Is this good to go?

**Comment by lance** [comment]
> Is this good to go?

Yes I think it is. Not sure why it has not landed.

File: pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml [hunk 6]
```diff
@@ -24,7 +24,16 @@ spec:
       description: The workspace containing the function project
   steps:
   - name: func-deploy
-    image: "registry.redhat.io/openshift-serverless-1/kn-plugin-func-rhel8:latest"
-    script: |
-      export FUNC_IMAGE="$(params.image)"
-      func deploy --verbose --build=false --push=false --path=$(params.path)
+    image: "registry.redhat.io/openshift-serverless-1/client-kn-rhel8@sha256:51e683fc3d668c39c6100512492e7fd7ca8f3f910886b4302960e825936d7cfb"
```
Comment (Kaustubh-pande) [inline pipelines/resources/tekton/task/func-deploy/0.1/func-deploy.yaml:27]: HI @jrangelramos New build sha
```suggestion
    image: "registry.redhat.io/openshift-serverless-1/client-kn-rhel8@sha256:8ef9539d11a14d19bdf2fa86704adcf9d8c59d613a4ecbb19c0b2475f33f4fa1" 
```

**Review by Kaustubh-pande — CHANGES_REQUESTED** [review]
